drivers/spi/spi_nrfx_spi.c

```diff
@@ -276,51 +273,32 @@ static int init_spi(const struct device *dev)
 		return -EBUSY;
 	}
 
-#ifdef CONFIG_PM_DEVICE
-	dev_data->pm_state = PM_DEVICE_STATE_ACTIVE;
-#endif
-
 	return 0;
 }
 
 #ifdef CONFIG_PM_DEVICE
 static int spi_nrfx_pm_control(const struct device *dev,
-				uint32_t ctrl_command,
-				enum pm_device_state *state)
+			       enum pm_device_state state)
 {
 	int ret = 0;
 	struct spi_nrfx_data *data = get_dev_data(dev);
 	const struct spi_nrfx_config *config = get_dev_config(dev);
 
-	if (ctrl_command == PM_DEVICE_STATE_SET) {
-		enum pm_device_state new_state = *state;
-
-		if (new_state != data->pm_state) {
-			switch (new_state) {
-			case PM_DEVICE_STATE_ACTIVE:
-				ret = init_spi(dev);
-				/* Force reconfiguration before next transfer */
-				data->ctx.config = NULL;
-				break;
-
-			case PM_DEVICE_STATE_LOW_POWER:
-			case PM_DEVICE_STATE_SUSPEND:
-			case PM_DEVICE_STATE_OFF:
-				if (data->pm_state == PM_DEVICE_STATE_ACTIVE) {
-					nrfx_spi_uninit(&config->spi);
-				}
-				break;
-
-			default:
-				ret = -ENOTSUP;
-			}
-			if (!ret) {
-				data->pm_state = new_state;
-			}
-		}
-	} else {
-		__ASSERT_NO_MSG(ctrl_command == PM_DEVICE_STATE_GET);
-		*state = data->pm_state;
+	switch (state) {
+	case PM_DEVICE_STATE_ACTIVE:
+		ret = init_spi(dev);
+		/* Force reconfiguration before next transfer */
+		data->ctx.config = NULL;
+		break;
+
+	case PM_DEVICE_STATE_LOW_POWER:
+	case PM_DEVICE_STATE_SUSPEND:
+	case PM_DEVICE_STATE_OFF:
+		nrfx_spi_uninit(&config->spi);
```

Comment:
you are unconditionally calling `nrfx_spi_uninit` while the original code just call it if the current state is `ACTIVE`.

Reply:
this is precisely the purpose of this PR: remove this redundant logic. There is no need to do this check, since the device will be in active state if it reaches this point.

Reply:
One calls `pm_device_state_set(dev, PM_DEVICE_STATE_LOW_POWER);` if the device is ACTIVE it will result in this code calling `nrfx_spi_uninit(&config->spi);`. Now the device is `PM_DEVICE_STATE_LOW_POWER`.

Now, if we do: `pm_device_state_set(dev, PM_DEVICE_STATE_OFF);` this function will end up calling  `nrfx_spim_uninit(&config->spim);` again right ? While in the original implementation this was only called if the device was ACTIVE.

Reply:
Indeed, this is a PM framework problem because we have poorly defined states and transitions, leading to this sort of issues. In a first place, `pm_device_state_set` needs to be removed from the public API space as we all agree it is not the proper way to handle device PM. Second, most (if not all) devices have this sort of fall-through to handle all "low power states", which is wrong. The possible state transitions should be limited, e.g. as Windows does https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/device-power-states

I have just opened an issue detailing all these problems https://github.com/zephyrproject-rtos/zephyr/issues/37056

#36125 goes further simplifying most of these problems, plus introducing the actions concept, which helps to prevent this sort of problems.

Reply:
Agree, but we should stop break everything thing while we don't have proper solutions. I like the idea of having defined transitions so devices don't need to check it by themselves. But even considering what is done by windows in https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/device-power-states, it is possible to go from D3Hot to D3Cold but it is not defined that all devices should support it.

So lets do an exercise, imagine this devices maps `PM_DEVICE_STATE_SUSPEND` to `D3Hot` and `PM_DEVICE_STATE_OFF` to `D3Cold`. In this scenario the transition between `PM_DEVICE_STATE_SUSPEND` to `PM_DEVICE_STATE_OFF` is allowed from the subsystem perspective, but this driver does not support it. It just change the state if the current state is `ACTIVE`.  With your patch this transition will happen. 

That is wrong ? I don't know, but I don't like change device drivers behavior in this massive changes. I vote to keep their behavior and after this being merged address it in another pr, that is smaller and will get attention from the proper maintainer. We *can't* assume we know everything about all devices.

Reply:
Btw, that is my only concern in this pr right now, everything else is great for me.

Reply:
I think the problem is that most drivers started following the pattern below because there is no clear guidance on what low power states mean as of today, so everything that sounded like "low power" was put into the same basket:

```c
	case PM_DEVICE_STATE_LOW_POWER:
	case PM_DEVICE_STATE_SUSPEND:
	case PM_DEVICE_STATE_OFF:
```

This design pattern has some problems. For example, suspend and off should not perform the same action since one will keep context while the other won't. There are actually many other problems that this PR is not fixing, see e.g.

```c
		if (*state == PM_DEVICE_STATE_ACTIVE) {
			ret = st7735r_exit_sleep(data);
			if (ret < 0) {
				return ret;
			}
			data->pm_state = PM_DEVICE_STATE_ACTIVE;
		} else {
			ret = st7735r_enter_sleep(data);
			if (ret < 0) {
				return ret;
			}
			data->pm_state = PM_DEVICE_STATE_LOW_POWER;
		}
```

In this case things are even worse, since any state `!= PM_DEVICE_STATE_ACTIVE` will call the sleep code and leave the device in `PM_DEVICE_STATE_LOW_POWER`, even if another state was requested.

Note that #36125 RFC fixes most (if not all) these problems in later patches by removing these bad patterns. But still, I think that until #37056 is solved, we won't have a final/proper solution.

Reply:
For reference, the final tentative code proposed in the RFC for the mentioned drivers:

https://github.com/zephyrproject-rtos/zephyr/blob/cebb50a54e375f1e5c7af96ae38d1602a70c2531/drivers/spi/spi_nrfx_spi.c#L280-L303

https://github.com/zephyrproject-rtos/zephyr/blob/cebb50a54e375f1e5c7af96ae38d1602a70c2531/drivers/display/display_st7735r.c#L510-L529